Delete duplicated OTLP Exporter tests, move them to the mixin unit test. Fix broken shutdown unit test.

DylanRussell opened the pull request:
# Description

Move all OTLP exporter tests that are testing the underlying behavior in the mixin to the mixin unit tests, 
instead of having them specified multiple times in the Metric/Log/Trace exporters.

Fix the shutdown tests which were flaky, so that they just test whether a pending export call completes or not. 

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

This change is just a refactor of the unit tests, to delete duplicate tests and fix a broken test.. No functionality is changing.

# Does This PR Require a Contrib Repo Change?
- [X] No.

# Checklist:

- [X] Followed the style guidelines of this project
- [ ] Changelogs have been updated
- I don't think a changelog update is necessary for this, but I can add one if needed.
- [X] Unit tests have been added
- [ ] Documentation has been updated

aabmass commented on exporter/opentelemetry-exporter-otlp-proto-grpc/src/opentelemetry/exporter/otlp/proto/grpc/exporter.py:
```diff
@@ -289,11 +289,11 @@ def _export(
         # exponentially. Once delay is greater than max_value, the yielded
         # value will remain constant.
-        for delay in _create_exp_backoff_generator(max_value=max_value):
-            if delay == max_value or self._shutdown:
```
I think the lock was not held here before on purpose, so that shutdown can be set and it will exit the export loop early.

There is also another problem where shutdown gets blocked on the sleep at L337 (https://github.com/open-telemetry/opentelemetry-python/issues/3309 and PR we never reviewed https://github.com/open-telemetry/opentelemetry-python/pull/3764/files), but at least this way it won't block for more than max_value seconds. We can discuss that later I suppose.

DylanRussell replied:
It sounds like we want a Condition and not a Lock: https://docs.python.org/3/library/threading.html#condition-objects

When the `export` call gives up the lock, there is no guarantee that other threads waiting for the lock will acquire it. See https://stackoverflow.com/questions/65905098/python-thread-not-releasing-lock

aabmass commented on exporter/opentelemetry-exporter-otlp-proto-grpc/src/opentelemetry/exporter/otlp/proto/grpc/exporter.py:
```diff
@@ -352,15 +352,14 @@ def _export(
         self._export_lock.acquire(timeout=timeout_millis / 1e3)
         self._shutdown = True
         self._channel.close()
-        self._export_lock.release()
```
Won't this leave the lock permanently acquired? I guess it's about to GC at this point anyway but still feels a little wrong.

See also https://github.com/open-telemetry/opentelemetry-python/issues/3398

DylanRussell replied:
Note that calling `acquire` with `timeout` means we proceed after `timeout` regardless of whether we acquired the lock: https://docs.python.org/3/library/threading.html#threading.Lock.acquire 

So we could change this to release the lock depending on if we acquired the lock or not. But it sounds like we want to use a conditional not a lock (see my other comment), WDYT ?

DylanRussell replied:
Resolving now that  I moved this stuff out of this PR

DylanRussell commented on exporter/opentelemetry-exporter-otlp-proto-grpc/src/opentelemetry/exporter/otlp/proto/grpc/exporter.py:
```diff
@@ -352,15 +352,14 @@ def _export(
 
         return self._result.FAILURE
 
-    def shutdown(self, timeout_millis: float = 30_000, **kwargs) -> None:
```
Ah ok, added it back. Shall I add it to the exporter classes then ?

xrmx commented:
@DylanRussell have you seen https://github.com/open-telemetry/opentelemetry-python/pull/4490?

DylanRussell commented:
Ah I had not seen that ! Seems like a couple of us noticed the same thing.

It sounds like the behavior that what we want is for `Shutdown()` to not interrupt any in progress export RPC, but also for it to prevent any additional retries and also we would like `Shutdown()` to interrupt the `sleep` call in `export` if possible.

I'm thinking we can use threading events (https://docs.python.org/3/library/threading.html#event-objects) to accomplish all these things. We use one event for `export` to communicate to `shutdown` that an RPC is in progress, and we use another event for `shutdown` to communicate to `export` that shutdown is happening.

Check out my latest commit and let me know what you think !

DylanRussell commented:
I reverted all changes to the exporter shutdown logic from this PR. So this PR only touches tests now, and is just about deleting duplicate tests and fixing the existing broken shutdown test.

I moved my proposed shutdown changes to a new pr: https://github.com/open-telemetry/opentelemetry-python/pull/4511

aabmass commented on exporter/opentelemetry-exporter-otlp-proto-grpc/tests/test_otlp_exporter_mixin.py:
```diff
@@ -26,20 +25,188 @@
+        args=(),
+        kwargs=None,
+    ):
+        threading.Thread.__init__(self, group, target, name, args, kwargs)
```
```suggestion
        super().__init__(self, group, target, name, args, kwargs)
```

DylanRussell replied:
Nice. Done

aabmass commented on exporter/opentelemetry-exporter-otlp-proto-grpc/tests/test_otlp_exporter_mixin.py:
```diff
@@ -26,20 +25,188 @@
+        self.server.add_insecure_port("127.0.0.1:4317")
+
+        self.server.start()
+        # The below tests use the SpanExporter and Spans, but are testing the
+        # underlying behavior in the mixin. A MetricExporter or LogExporter could
```
Consider making a minimal implementation of Span exporter in this file that uses the mixin for clarity and to isolate the behavior

DylanRussell replied:
Good idea, done

aabmass commented on exporter/opentelemetry-exporter-otlp-proto-grpc/tests/test_otlp_exporter_mixin.py:
```diff
@@ -26,20 +25,188 @@
+            self._return = self._target(*self._args, **self._kwargs)  # type: ignore
+
+    def join(self, *args):  # type: ignore
+        threading.Thread.join(self, *args)
```
```suggestion
        super().join(self, *args)
```

DylanRussell replied:
Nice. Done

aabmass commented on exporter/opentelemetry-exporter-otlp-proto-grpc/tests/test_otlp_exporter_mixin.py:
```diff
@@ -64,147 +231,173 @@ def test_environ_to_compression(self):
+        # Wait a sec for exporter to get lock and make export call.
+        time.sleep(1)
+        # pylint: disable=protected-access
+        self.assertTrue(self.exporter._export_lock.locked())
```
How do you klnow that the export lock will still be held?

DylanRussell replied:
Because the server is setup to wait 5 seconds before responding. That is within the export lock block.

aabmass replied:
If you can make these test work without sleeping, it would be awesome. The sleeping slows things down but also we've seen flakes with PyPy tests from this kind of thing. If you can't get that working, consider lowering the sleep times so the test runs faster.

`Barrier` or `Semaphore` might be handy

aabmass reviewed (commented):
Did you mostly rewrite the tests in `exporter/opentelemetry-exporter-otlp-proto-grpc/tests/test_otlp_exporter_mixin` or does it copy from the old tests that were deleted here?

DylanRussell commented:
I copied the existing tests. The only tests I changed were the shutdown tests (`test_shutdown.*`) because they were flaky / broken

aabmass commented on exporter/opentelemetry-exporter-otlp-proto-grpc/tests/test_otlp_exporter_mixin.py:
```diff
@@ -64,147 +281,171 @@ def test_environ_to_compression(self):
+        # Shutdown waits 30 seconds for a pending Export call to finish.
+        # A 5 second delay is added, so it's expected that export will finish.
+        add_TraceServiceServicer_to_server(
+            TraceServiceServicerWithExportParams(StatusCode.OK, 5),
```
nit consider passing the args as named params, it's hard to tell what `5` is in this case.

aabmass reviewed (approved):
If you can make the tests work without sleeping, it would be nice. If not, can you try to shorten the sleep durations a bit?

DylanRussell commented on exporter/opentelemetry-exporter-otlp-proto-grpc/tests/test_otlp_exporter_mixin.py:
```diff
@@ -26,20 +26,238 @@
+        target=None,
+        args=(),
+    ):
+        super().__init__(None, target, None, args, None)
```
Nice, yes.

xrmx commented on exporter/opentelemetry-exporter-otlp-proto-grpc/tests/test_otlp_exporter_mixin.py:
```diff
@@ -26,20 +26,230 @@
+    def tearDown(self):
+        self.server.stop(None)
+
+    # pylint: disable=no-self-use
```
```suggestion
```

DylanRussell replied:
Done

xrmx commented on exporter/opentelemetry-exporter-otlp-proto-grpc/src/opentelemetry/exporter/otlp/proto/grpc/trace_exporter/__init__.py:
```diff
@@ -142,8 +142,8 @@ def _translate_data(
 
-    def shutdown(self) -> None:
-        OTLPExporterMixin.shutdown(self)
+    def shutdown(self, timeout_millis: float = 30_000) -> None:
```
Is this a fix?

DylanRussell replied:
Yeah let me remove this from the PR. I think we should do it but doesn't belong in this PR.

xrmx reviewed (approved):
I like the reduction of work done, hopefully we are still confident that the exporter behaves like the mixin. A couple of nits but LGTM.

xrmx commented:
@jomcgi care to take a look at this too please?

jomcgi commented on exporter/opentelemetry-exporter-otlp-proto-grpc/tests/test_otlp_exporter_mixin.py:
```diff
@@ -26,20 +26,213 @@
+        self.optional_export_sleep = optional_export_sleep
+        self.optional_export_retry_millis = optional_export_retry_millis
+
+    # pylint: disable=invalid-name,unused-argument,no-self-use
```
nit: unnecessary? if the self use is recognized by pylint
```suggestion
    # pylint: disable=invalid-name,unused-argument
```

DylanRussell replied:
Yeah you're right it's not needed. Removed

jomcgi commented on exporter/opentelemetry-exporter-otlp-proto-grpc/tests/test_otlp_exporter_mixin.py:
```diff
@@ -26,20 +26,213 @@
     RetryInfo,
 )
-from grpc import Compression
+from grpc import Compression, server
```
nit: We're importing `StatusCode` from `grpc` via `opentelemetry.exporter.otlp.proto.grpc.exporter` instead of directly.
```suggestion
from grpc import Compression, server, StatusCode
```

DylanRussell replied:
Nice, done

jomcgi commented on exporter/opentelemetry-exporter-otlp-proto-grpc/tests/test_otlp_exporter_mixin.py:
```diff
@@ -26,20 +26,213 @@
+class TraceServiceServicerWithExportParams(TraceServiceServicer):
+    def __init__(
+        self,
+        export_result,
+        optional_export_sleep=None,
+        optional_export_retry_millis=None,
```
nit: typing
```suggestion
        export_result: StatusCode,
        optional_export_sleep: Optional[float] = None,
        optional_export_retry_millis: Optional[float] = None,
```

DylanRussell replied:
Added

jomcgi reviewed (approved):
Looks good to me! left some small nits.

emdneto commented on exporter/opentelemetry-exporter-otlp-proto-grpc/tests/logs/test_otlp_logs_exporter.py:
```diff
@@ -296,145 +226,6 @@ def test_env_variables_with_only_certificate(
-        "opentelemetry.exporter.otlp.proto.grpc._log_exporter.OTLPLogExporter._stub"
-    )
-    # pylint: disable=unused-argument
-    def test_no_credentials_error(
```
no needed anymore then?

DylanRussell replied:
I think I thought this test was meaningless because there is no "error". But I've added it back now, just in the mixin test file

emdneto commented on exporter/opentelemetry-exporter-otlp-proto-grpc/src/opentelemetry/exporter/otlp/proto/grpc/exporter.py:
```diff
@@ -369,4 +369,4 @@ def _exporting(self) -> str:
         warning messages.
         """
-        pass
+        pass
```
tox -e precommit should fix precommit ci

DylanRussell replied:
Done thanks

DylanRussell commented:
I think I need someone to add the Skip Changelog label
